ros2topic/ros2topic/verb/hz.py

```diff
@@ -0,0 +1,337 @@
+# POSSIBILITY OF SUCH DAMAGE.
+
+# This file is originally from:
+# https://github.com/ros/ros_comm/blob/6e5016f4b2266d8a60c9a1e163c4928b8fc7115e/tools/rostopic/src/rostopic/__init__.py
+
+from argparse import ArgumentTypeError
+
+import functools
+import importlib
+import math
+
+from time import sleep
+
+import rclpy
+
+from rclpy.clock import Clock
+from rclpy.clock import ClockType
+from rclpy.expand_topic_name import expand_topic_name
+from rclpy.qos import qos_profile_sensor_data
+from rclpy.validate_full_topic_name import validate_full_topic_name
+from ros2cli.node.direct import DirectNode
+from ros2topic.api import get_topic_names_and_types
+from ros2topic.api import TopicNameCompleter
+from ros2topic.verb import VerbExtension
+
+DEFAULT_WINDOW_SIZE = 10000
+
+
+def unsigned_int(string):
+    try:
+        value = int(string)
+    except ValueError:
+        value = -1
+    if value < 0:
+        raise ArgumentTypeError('value must be non-negative integer')
+    return value
+
+
+class HzVerb(VerbExtension):
+    """Print the average publishing rate to screen."""
+
+    def add_arguments(self, parser, cli_name):
+        arg = parser.add_argument(
+            'topic_name',
+            help="Name of the ROS topic to listen to (e.g. '/chatter')")
+        arg.completer = TopicNameCompleter(
+            include_hidden_topics_key='include_hidden_topics')
+        parser.add_argument(
+            '--window', '-w',
+            dest='window_size', type=unsigned_int, default=DEFAULT_WINDOW_SIZE,
+            help='window size, in # of messages, for calculating rate, '
+                 'string to (default: %d)' % DEFAULT_WINDOW_SIZE, metavar='WINDOW')
+        parser.add_argument(
+            '--filter',
+            dest='filter_expr', default=None,
+            help='only measure messages matching the specified Python expression', metavar='EXPR')
+        parser.add_argument(
+            '--wall-time',
+            dest='use_wtime', default=False, action='store_true',
+            help='calculates rate using wall time which can be helpful'
+                 ' when clock is not published during simulation')
+
+    def main(self, *, args):
+        return main(args)
+
+
+def main(args):
+    topic = args.topic_name
+    if args.filter_expr:
+        def expr_eval(expr):
+            def eval_fn(m):
+                return eval(expr)
+            return eval_fn
+        filter_expr = expr_eval(args.filter_expr)
+    else:
+        filter_expr = None
+
+    with DirectNode(args) as node:
+        _rostopic_hz(node.node, topic, window_size=args.window_size, filter_expr=filter_expr,
+                     use_wtime=args.use_wtime)
+
+
+class ROSTopicHz(object):
+    """ROSTopicHz receives messages for a topic and computes frequency."""
+
+    def __init__(self, node, window_size, filter_expr=None, use_wtime=False):
+        import threading
+        from collections import defaultdict
```

Comment:
Nit; looking elsewhere in ros2cli, it looks like we generally hoist `defaultdict` to the top of the file: https://github.com/ros2/ros2cli/blob/master/ros2cli/ros2cli/entry_points.py#L16 .  Mostly a matter of style/taste, just trying to be consistent here.

Reply:
fixed in
0ba39d3